src/plugins/admin/admin.plugin.ts

```diff
@@ -0,0 +1,172 @@
+      maxAge: ADMIN_SESSION_EXPIRATION_IN_SECONDS,
+    },
+    expiry: ADMIN_SESSION_EXPIRATION_IN_SECONDS,
+  });
+  // declare a new authenticator so that passport for admins does not interfere with the user passport strategies
+  // also set the user property to `admin` so that it will be available on `request.admin`
+  const adminPassport = new Authenticator({ userProperty: 'admin' });
+  await fastify.register(adminPassport.initialize());
+  await fastify.register(adminPassport.secureSession());
+
+  adminPassport.use(
+    GITHUB_OAUTH_STRATEGY,
+    new GitHubStrategy(
+      {
+        clientID: GITHUB_CLIENT_ID,
+        clientSecret: GITHUB_CLIENT_SECRET,
+        callbackURL: `${PUBLIC_URL.origin}/admin/auth/github/callback`,
+      },
+      async (accessToken: string, refreshToken: string, profile: GitHubProfile, done) => {
+        const { username, id: githubId } = profile;
+        if (!githubId) {
+          throw new MissingGithubId();
+        }
+        if (!username) {
+          throw new MissingGithubUsername();
+        }
+        // only allow users that are present in the admin table by their username
+        if (await adminRepository.isAdmin(db, githubId)) {
+          console.debug('user is an allowed admin');
+          // update info stored in the table
+          await adminRepository.update(db, { githubId, githubName: username });
+          // You can add admin checks here
+          return done(null, profile);
+        }
+        console.debug('user is not an allowed admin', profile);
+        return done(new NotAnAuthorizedAdmin());
+      },
+    ),
+  );
+
+  adminPassport.registerUserSerializer(async (user: GitHubProfile, req: AdminRequest) => {
+    req.log.info(user);
+    return user.id;
+  });
+  adminPassport.registerUserDeserializer(
+    async (uuid: string, req: AdminRequest): Promise<AdminUser | undefined> => {
+      req.log.info({ uuid }, 'Deserialize admin github id');
+
+      const admin = await adminRepository.get(db, uuid);
+      req.log.info({ admin }, 'Resolved github id to admin user');
+
+      return admin;
+    },
+  );
+
+  fastify.get(
+    '/admin/auth/github',
+    {
+      preValidation: adminPassport.authenticate(GITHUB_OAUTH_STRATEGY, { scope: ['user:email'] }),
+    },
+    async (_request: AdminRequest, _reply) => {},
+  );
+
+  fastify.get(
+    '/admin/auth/github/callback',
+    {
+      preValidation: adminPassport.authenticate(GITHUB_OAUTH_STRATEGY, {
+        failureRedirect: '/admin/login',
+      }),
+    },
+    async (_request: AdminRequest, reply) => {
+      reply.redirect('/admin');
+    },
+  );
+
+  // login page when user is not authenticated
+  fastify.get('/admin/login', async (_request, reply) => {
+    reply.type('text/html').send('<a href="/admin/auth/github">Login with GitHub</a>');
+  });
+
+  // create a scope where if the user is not authenticated they get redirected to the login page
+  fastify.register(
+    fp(async (authenticatedAdmin) => {
```

Comment:
Do you need `fp`?

Reply:
I do not need `fp` in this instance, because the upper scope does not need access to what will be defined in this plugin. We always get access to what is defined in the parent which is enough for this case.